File: CrypticPassword_1.5.0/android/app/src/main/java/com/anoop/crypticpassword/CrypticPasswordAlgorithm.java
package com.anoop.crypticpassword;

public final class CrypticPasswordAlgorithm {
    public enum Version {
        VERSION_1_0("version1_0 / seed1"),
        VERSION_1_1("version1_1 / seed1"),
        VERSION_2_0("version2_0 / seed2"),
        VERSION_2_1("version2_1 / seed2");

        public final String label;
        Version(String label) { this.label = label; }
    }

    public static final int MATRIX_COUNT = 3;
    public static final int MATRIX_ROWS = 6;
    public static final int MATRIX_COLUMNS = 6;
    public static final int MATRIX_SIZE = 36;
    public static final int TOTAL_CELLS = 108;
    public static final int PIN_MINIMUM = 1;
    public static final int PIN_MAXIMUM = 99999;

    public static final String SEED1 = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyz!$'*+,-./:;<=>?^_`~$'*+,-./:;<=>?^_`";
    public static final String SEED2 = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz`~!@#$%^&*()_+-={}|[]\\\\:\\\";'<>?,./^&*()_+-={}|[]\\\";";

    private CrypticPasswordAlgorithm() {}

    public static int normalizePin(int pin) {
        if (pin < PIN_MINIMUM) return PIN_MINIMUM;
        if (pin > PIN_MAXIMUM) return PIN_MAXIMUM;
        return pin;
    }

    private static String seedFor(Version version) {
        switch (version) {
            case VERSION_2_0:
            case VERSION_2_1:
                return SEED2;
            case VERSION_1_0:
            case VERSION_1_1:
            default:
                return SEED1;
        }
    }

    private static void rotate(char[] data, int start, int count, int length) {
        if (length <= 0) return;
        int shift = count % length;
        if (shift <= 0) return;
        char[] temp = new char[length];
        System.arraycopy(data, start + shift, temp, 0, length - shift);
        System.arraycopy(data, start, temp, length - shift, shift);
        System.arraycopy(temp, 0, data, start, length);
    }

    private static void swap(char[] data, int base, int i, int j) {
        char tmp = data[base + i];
        data[base + i] = data[base + j];
        data[base + j] = tmp;
    }

    private static void shuffleSimple(char[] data, int pin) {
        final int n = TOTAL_CELLS;
        final int half = n / 2;
        int k = 1;
        for (int j = 0; j < pin; j++) {
            for (int i = 0; i < n / 4; i++) {
                if (k % 2 == 0) {
                    swap(data, 0, i, half - i - 1);
                    swap(data, half, i, half - i - 1);
                } else {
                    swap(data, 0, i, half - i - 2);
                    swap(data, half, i, half - i - 2);
                }
                k++;
            }
            rotate(data, 0, 13, half);
            rotate(data, half, 17, half);
            rotate(data, 0, 23, n);
        }
    }

    private static void shuffleEnhanced(char[] data, int pin) {
        final int n = TOTAL_CELLS;
        final int half = n / 2;
        final int[] random = {23, 13, 17, 7, 11};
        int k = 1;
        for (int j = 0; j < pin; j++) {
            for (int i = 0; i < n / 4; i++) {
                if (k % 2 == 0) {
                    swap(data, 0, i, half - i - 1);
                    swap(data, random[3], i, half - i - 1 - random[2]);
                    swap(data, half, i, half - i - 1);
                    swap(data, random[4], i, half - i - 1 - random[4]);
                } else {
                    swap(data, 0, i, half - i - 2);
                    swap(data, random[3], i, half - i - 2 - random[3]);
                    swap(data, half, i, half - i - 2);
                    swap(data, random[4], i, half - i - 2 - random[4]);
                }
                swap(data, 0, i, random[1] + i);
                swap(data, half, i, random[2] + i);
                k++;
            }
            rotate(data, 0, random[1], half);
            rotate(data, half, random[2], half);
            rotate(data, 0, random[0], n - random[0]);
            rotate(data, 0, random[3], n - random[3]);
            rotate(data, random[3], random[4], half);
        }
    }

    public static String shuffle(Version version, int rawPin) {
        int pin = normalizePin(rawPin);
        String seed = seedFor(version);
        char[] data = seed.substring(0, TOTAL_CELLS).toCharArray();
        switch (version) {
            case VERSION_1_1:
            case VERSION_2_1:
                shuffleEnhanced(data, pin);
                break;
            case VERSION_1_0:
            case VERSION_2_0:
            default:
                shuffleSimple(data, pin);
                break;
        }
        return new String(data, 0, TOTAL_CELLS);
    }
}

File: CrypticPassword_1.5.0/ios/src/CrypticPasswordAlgorithm.swift
import Foundation

final class CrypticPasswordAlgorithm {
    enum Version: Int, CaseIterable {
        case version1_0 = 0
        case version1_1 = 1
        case version2_0 = 2
        case version2_1 = 3
    }

    static let matrixCount = 3
    static let matrixRows = 6
    static let matrixColumns = 6
    static let matrixSize = 36
    static let totalCells = 108
    static let pinMinimum = 1
    static let pinMaximum = 99_999

    static let algorithmLabels = [
        "version1_0 / seed1",
        "version1_1 / seed1",
        "version2_0 / seed2",
        "version2_1 / seed2"
    ]

    static let seed1 = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyz!$'*+,-./:;<=>?^_`~$'*+,-./:;<=>?^_`"
    static let seed2Full = #"0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz`~!@#$%^&*()_+-={}|[]\\:\";'<>?,./^&*()_+-={}|[]\";"#

    static func normalize(pin: Int) -> Int {
        return min(max(pin, pinMinimum), pinMaximum)
    }

    static func normalized(version rawValue: Int) -> Version {
        return Version(rawValue: rawValue) ?? .version1_0
    }

    private static func seed(for version: Version) -> String {
        switch version {
        case .version1_0, .version1_1:
            return seed1
        case .version2_0, .version2_1:
            return seed2Full
        }
    }

    private static func rotate(_ data: inout [Character], start: Int, count: Int, length: Int) {
        guard length > 0 else { return }
        let shift = count % length
        guard shift > 0 else { return }
        let segment = Array(data[start..<(start + length)])
        let rotated = Array(segment[shift..<length]) + Array(segment[0..<shift])
        for i in 0..<length {
            data[start + i] = rotated[i]
        }
    }

    private static func swap(_ data: inout [Character], base: Int, i: Int, j: Int) {
        data.swapAt(base + i, base + j)
    }

    private static func shuffleSimple(_ data: inout [Character], pin: Int) {
        let n = totalCells
        let half = n / 2
        var k = 1

        for _ in 0..<pin {
            for i in 0..<(n / 4) {
                if k % 2 == 0 {
                    swap(&data, base: 0, i: i, j: half - i - 1)
                    swap(&data, base: half, i: i, j: half - i - 1)
                } else {
                    swap(&data, base: 0, i: i, j: half - i - 2)
                    swap(&data, base: half, i: i, j: half - i - 2)
                }
                k += 1
            }

            rotate(&data, start: 0, count: 13, length: half)
            rotate(&data, start: half, count: 17, length: half)
            rotate(&data, start: 0, count: 23, length: n)
        }
    }

    private static func shuffleEnhanced(_ data: inout [Character], pin: Int) {
        let n = totalCells
        let half = n / 2
        let random = [23, 13, 17, 7, 11]
        var k = 1

        for _ in 0..<pin {
            for i in 0..<(n / 4) {
                if k % 2 == 0 {
                    swap(&data, base: 0, i: i, j: half - i - 1)
                    swap(&data, base: random[3], i: i, j: half - i - 1 - random[2])
                    swap(&data, base: half, i: i, j: half - i - 1)
                    swap(&data, base: random[4], i: i, j: half - i - 1 - random[4])
                } else {
                    swap(&data, base: 0, i: i, j: half - i - 2)
                    swap(&data, base: random[3], i: i, j: half - i - 2 - random[3])
                    swap(&data, base: half, i: i, j: half - i - 2)
                    swap(&data, base: random[4], i: i, j: half - i - 2 - random[4])
                }
                swap(&data, base: 0, i: i, j: random[1] + i)
                swap(&data, base: half, i: i, j: random[2] + i)
                k += 1
            }

            rotate(&data, start: 0, count: random[1], length: half)
            rotate(&data, start: half, count: random[2], length: half)
            rotate(&data, start: 0, count: random[0], length: n - random[0])
            rotate(&data, start: 0, count: random[3], length: n - random[3])
            rotate(&data, start: random[3], count: random[4], length: half)
        }
    }

    static func shuffle(version: Version, pin rawPin: Int) -> String {
        let pin = normalize(pin: rawPin)
        var data = Array(seed(for: version).prefix(totalCells))

        switch version {
        case .version1_0, .version2_0:
            shuffleSimple(&data, pin: pin)
        case .version1_1, .version2_1:
            shuffleEnhanced(&data, pin: pin)
        }

        return String(data.prefix(totalCells))
    }
}
